src/main/java/com/odysseusinc/arachne/portal/service/impl/AuthenticationServiceImpl.java

```diff
@@ -0,0 +1,82 @@
+import com.odysseusinc.arachne.portal.model.factory.ArachneUserFactory;
+import com.odysseusinc.arachne.portal.model.security.ArachneUser;
+import com.odysseusinc.arachne.portal.service.AuthenticationService;
+import org.ohdsi.authenticator.model.UserInfo;
+import org.ohdsi.authenticator.service.authentication.Authenticator;
+import org.pac4j.core.credentials.UsernamePasswordCredentials;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
+import org.springframework.security.core.context.SecurityContextHolder;
+import org.springframework.security.core.userdetails.UserDetails;
+import org.springframework.security.core.userdetails.UserDetailsService;
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Transactional;
+
+@Service
+@Transactional(rollbackFor = Exception.class, readOnly = false)
+public class AuthenticationServiceImpl implements AuthenticationService {
+
+    protected Authenticator authenticator;
+    private UserInfoToUserConverter userConverter;
+    private UserDetailsService userDetailsService;
+
+    @Value("${security.method}")
+    private String authMethod;
+
+
+    public AuthenticationServiceImpl(Authenticator authenticator, UserInfoToUserConverter userConverter, UserDetailsService userDetailsService) {
+
+        this.authenticator = authenticator;
+        this.userConverter = userConverter;
+        this.userDetailsService = userDetailsService;
+    }
+
+    @Transactional
+    public String authenticateAndGetAuthToken(CommonAuthenticationRequest authenticationRequest) {
+
+        String username = authenticationRequest.getUsername();
+        String password = authenticationRequest.getPassword();
+        try {
+            UserInfo userInfo = authenticator.authenticate(
+                    authMethod,
+                    new UsernamePasswordCredentials(username, password)
+            );
+            authenticateBaseOnExternalUser(authenticationRequest, userInfo);
+            synchronizeExternalAndInternalUserAccounts(username);
+            authenticateBaseOnInternalUser(authenticationRequest, username);
+            return userInfo.getAuthenticationInfo().getToken();
+        } catch (Exception e) {
+            SecurityContextHolder.clearContext();
+            throw e;
+        }
+    }
+
+    public void authenticateBaseOnExternalUser(CommonAuthenticationRequest authenticationRequest, UserInfo userInfo) {
+
+        User user = userConverter.convert(userInfo);
+        ArachneUser arachneUser = ArachneUserFactory.create(user);
+        UsernamePasswordAuthenticationToken authentication =
+                new UsernamePasswordAuthenticationToken(arachneUser, authenticationRequest.getPassword(), ((UserDetails) arachneUser).getAuthorities());
+        authentication.setDetails(arachneUser);
+        SecurityContextHolder.getContext().setAuthentication(authentication);
```

Comment:
yes, this is a good question, let me explain:
The first step[1] get User info from authentication provider (e.g. git, rest etc) 
then [2]we create/update the user in our current DB.  

I know that theoretically user info from provider and user saved in DB should be the same, but I cannot guarantee that, and we can add/change some specific privileges during the second step. 
so eventually we should put the session based on the data from db. 
on the other hand, we cannot get rid of the authentification based on a user from an Authentication provider, because the second step should be secured as well.

Reply:
I slightly change the logic so it is not so confusing now.